### PR TITLE
feat: implement student lookup with validation and test coverage

### PR DESCRIPTION
**Pull request by Lokeshloks-1999**
#  Student Lookup Feature

## Summary

This PR introduces the `findStudentById`  function along with comprehensive test cases to support student lookup functionality based on ID. It enhances the robustness of the student data system by ensuring correct handling of valid, invalid, and edge cases.

---

##  What's Added

- **`findStudentById.ts`**:  function to retrieve a student by their ID from a provided student getter function.
- **Unit tests** for:
  - Valid ID returns correct student.
  - Invalid ID returns `undefined`.
  - Invalid getter (e.g., returns null) throws a descriptive error.

---

##  Test Coverage

- All major paths are tested including:
  - Successful student retrieval
  - ID not found scenario
  - Data source (getter) is invalid scenario

---

## Dependency Note

This PR **imports the `Email` value object**, which was implemented in a previous PR:  
`fix: add Email value object with validation and tests`

## GitHub Issues

- https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/issues/434
- https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/issues/435

### CODE REVIEW

**Comment by Lokeshloks-1999** [comment]
Hello professor,
I’ve implemented the StudentEmailWrapper component, which properly uses findStudentById to load student details. The StudentEmailWrapper acts as the controller logic component for StudentEmail. 
Please let me know if any further adjustments are needed.

**Comment by Lokeshloks-1999** [comment]
Hello Professor @achow-conestogac 

I wanted to let you know that I’ve added a route in the application for displaying student details by ID.

Specifically, I created a route at /students/:id that renders the StudentDetailsByID component. This component is now independent and shows the student’s details based on the URL parameter. For now, I’ve set up the studentGetter with empty data as a placeholder, so it can later be connected to the real data source without further routing changes.

This setup is important because in the upcoming milestones, the email component will rely on the same logic to fetch each student’s details dynamically by ID when composing emails. Setting this up early ensures the routing and data retrieval are consistent and will simplify integrating the email functionality later on.

Please let me know if you’d like me to adjust anything.

File: src/App.tsx
Comment (achow-conestogac) [inline src/App.tsx:1]: you have the same problem as @Shiva30jun and @Charan-devv so i will copy/paste the message here:

this usage does not make sense.

Your visualization is part of enrollment.  it should not be a route that you navigate to.

you need to put your changes into the appropriate place in the Enrollment Dashboard.

File: src/App.tsx [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
 import EnrollmentDashboard from './Enrollment/Dashboard/Dashboard';
 import EvaluationCard from '../src/Dashboard/EvaluationCard';
+import StudentDetailsByID from './Enrollment/Student/Pages/StudentEmail/StudentDetailsByID';
 function App() {
   return (
     <BrowserRouter>
       <Routes>
-        <Route path="/enrollment" element={<EnrollmentDashboard />} />
+        <Route path="/enrollment" element={<EnrollmentDashboard />} >
+          <Route path="students/id" element={<StudentDetailsByID studentGetter={() => []}/>}/>
```
Comment (achow-conestogac) [inline src/App.tsx:13]: why does the `studentGetter` do nothing.

File: src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.test.tsx
Comment (achow-conestogac) [inline src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.test.tsx:1]: there are more things you can test for.

for one, there's more than 2 elements that you are displaying. 

there are also a bunch of error states that you have not tested.

File: src/Enrollment/Student/studentData/findStudentById.test.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import { findStudentById } from "./findStudentById";
+
+import { Student } from "./studentTypes";
+
+
```
Comment (achow-conestogac) [inline src/Enrollment/Student/studentData/findStudentById.test.ts:5]: fix formatting.

File: src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  let student;
+  try {
+    student = findStudentById(id, studentGetter);
+  } catch (err) {
+    return (
+      <div className="p-8 text-red-600 text-lg">
+        Error: {(err as Error).message}
+      </div>
+    );
+  }
```
Comment (achow-conestogac) [inline src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.tsx:30]: dont do this. best practice is to assign the variable as soon as it is used.

your `findStudentById` can return `undefined`. check for that instead.

```suggestion
  const student = findStudentById(id, studentGetter);
  
  if (!student) { // explanation: ! means "not" so if you do undefined is considered false so the opposite of false is true.  so if undefined, then you will end up inside this if block which will return 
    return (
      <div className="p-8 text-red-600 text-lg">
        Error: {(err as Error).message} // you need to fix this error thing here. this never worked as your function never throws an error.  this shows that TDD is not followed.
      </div>
    );
  }
```

Comment (Lokeshloks-1999) [reply src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.tsx]: I’ve updated the code by removing the unnecessary try-catch block and directly assigning the student variable when it’s needed. I also cleaned up the logic to handle the undefined case without relying on error messages. The component now follows best practices and is much cleaner.

File: src/App.tsx (hunk 1 above)
Comment (achow-conestogac) [reply src/App.tsx]: there's a problem with your structure then.

this is the "routing page".  you should not be trying to get data on this page.

so this means that the `EnrollmentDashboard` should have the data and should be the one giving that to you.

meaning, you should not be defining this as a route.  Rather on the `EnrollmentDashboard`, you need to pass in the correct data when this page is supposed to be displayed.

Comment (Lokeshloks-1999) [reply src/App.tsx]: @achow-conestogac 
Thank you again.

You're right — the routing layer shouldn't manage data logic directly. Initially, I exposed StudentDetailsByID as a route in App.tsx for early development and testing purposes, but you're correct that this doesn't align with best practices.

To clarify, StudentDetailsByID is a presentation component that displays detailed information for a student based on their ID. While it will eventually be reused across modules like StudentEmail, it’s not intended to be a standalone route at this point. Instead, it will be conditionally rendered inside components like StudentEmail, which does belong within the EnrollmentDashboard.

Given that, I’ll:
- Remove the route for StudentDetailsByID from App.tsx
- Ensure that EnrollmentDashboard handles routing to modules like StudentEmail
- Render StudentDetailsByID within StudentEmail using shared context for student data

This should align better with our structure and avoid premature routing setup.

Please let me know if you'd like me to approach this differently.

File: src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      <div className="p-8 text-red-600 text-lg">
+        No student ID provided in URL.
+      </div>
+    );
+  }
+  const student = findStudentById(id, studentGetter);
+  if (!student) {
+    return (
+      <div className="p-8 text-red-600 text-lg">
+        Student not found for ID = {id}
```
[the remark this answers is not in the document]
Comment (Lokeshloks-1999) [reply src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.tsx]: I have removed the id from the message for privacy and security reasons.

**Comment by Lokeshloks-1999** [comment]
Hello professor, I've updated StudentdetailsbyId to handle all null checks
- When Id not present
- When searching student information, show loading
- When student information not found for the given Id in the route.

I also moved the getterfunction from the StudentdetailbyId route in App.tsx to the component itself, which will be replaced during integration.

File: src/Enrollment/Student/Pages/StudentList/StudentList.tsx [hunk 5]
```diff
@@ -54,6 +58,7 @@ const StudentList: React.FC = () => {
             <th className="px-4 py-3 text-right">Actions</th>
           </tr>
         </thead>
+        {/* wire the navigateToStudentDetails to onClick of the student name in the row */}
```
[the remark this answers is not in the document]
Comment (Lokeshloks-1999) [reply src/Enrollment/Student/Pages/StudentList/StudentList.tsx]: Did the cleanup of comment.

File: src/Enrollment/Student/Pages/StudentList/StudentList.tsx
[the remark this answers is not in the document]
Comment (Lokeshloks-1999) [reply src/Enrollment/Student/Pages/StudentList/StudentList.tsx:1]: Removed our code from studentlist, so no tests are required.

**Comment by Lokeshloks-1999** [comment]
Hello professor, I have updated my code accordingly. Please review it.

File: src/App.tsx
Comment (achow-conestogac) [reply src/App.tsx:1]: hello lokesh, you have not address this comment.

I am telling you that you should NOT be using routing but to implement your code directly into the dashboard.

Comment (Lokeshloks-1999) [reply src/App.tsx:1]: Hello Professor, Thank you for the clarification. I now understand that the expectation is to avoid routing altogether and instead integrate the visualization directly into the Enrollment Dashboard.

I had initially followed the structure from Shiva’s PR(#573), which is why I approached it the same way. In hindsight, I realize this was more of a planning and integration oversight on our part during the initial stages. I did explore different ways to keep the visualization code independent of the StudentList, but due to the tight coupling, it was quite challenging.

Once Shiva’s code is approved, I will align my changes accordingly and refactor the implementation to properly integrate it within the dashboard component, as per your direction.
